sympy/algebras/quaternion.py

```diff
@@ -0,0 +1,540 @@
+                else:
+                    return Mul(q1, q2)
+            else:
+                return Quaternion(q1 * q2.a, q1 * q2.b, q1 * q2.c, q1 * q2.d)
+
+
+        # If q2 is a number or a sympy expression instead of a quaternion
+        if not isinstance(q2, Quaternion):
+            if q1.real_field:
+                if q2.is_complex:
+                    return q1 * Quaternion(re(q2), im(q2), 0, 0)
+                else:
+                    return Mul(q1, q2)
+            else:
+                return Quaternion(q2 * q1.a, q2 * q1.b, q2 * q1.c, q2 * q1.d)
+
+        return Quaternion(-q1.b*q2.b - q1.c*q2.c - q1.d*q2.d + q1.a*q2.a,
+                          q1.b*q2.a + q1.c*q2.d - q1.d*q2.c + q1.a*q2.b,
+                          -q1.b*q2.d + q1.c*q2.a + q1.d*q2.b + q1.a*q2.c,
+                          q1.b*q2.c - q1.c*q2.b + q1.d*q2.a + q1.a * q2.d)
+
+    def _eval_conjugate(self):
+        """Returns the conjugate of the quaternion."""
+        q = self
+        return Quaternion(q.a, -q.b, -q.c, -q.d)
+
+    def norm(self):
+        """Returns the norm of the quaternion."""
+        q = self
+        # trigsimp is used to simplify sin(x)^2 + cos(x)^2 (these terms
+        # arise when from_axis_angle is used).
+        return sqrt(trigsimp(q.a**2 + q.b**2 + q.c**2 + q.d**2))
+
+    def normalize(self):
+        """Returns the normalized form of the quaternion."""
+        q = self
+        return q * (1/q.norm())
+
+    def inverse(self):
+        """Returns the inverse of the quaternion."""
+        q = self
+        return conjugate(q) * (1/q.norm()**2)
+
+    def pow(self, p):
+        """Finds the pth power of the quaternion.
+        Returns the inverse if p = -1.
+
+        Example
+        ========
+
+        >>> from sympy.algebras.quaternion import Quaternion
+        >>> q = Quaternion(1, 2, 3, 4)
+        >>> q.pow(4)
+        668 + (-224)*i + (-336)*j + (-448)*k
+        """
+        q = self
+        if p == -1:
+            return q.inverse()
+        res = 1
+        while p > 0:
+            if p & 1:
+                res = q * res
+
+            p = p >> 1
+            q = q * q
+
+        return res
+
+    def exp(self):
+        """Returns the exponential of q (e^q).
+
+        Example
+        ========
+
+        >>> from sympy.algebras.quaternion import Quaternion
+        >>> q = Quaternion(1, 2, 3, 4)
+        >>> q.exp()
+        E*cos(sqrt(29))
+        + 2*sqrt(29)*E*sin(sqrt(29))/29*i
+        + 3*sqrt(29)*E*sin(sqrt(29))/29*j
+        + 4*sqrt(29)*E*sin(sqrt(29))/29*k
+        """
+        # exp(q) = e^a(cos||v|| + v/||v||*sin||v||)
+        q = self
+        vector_norm = sqrt(q.b**2 + q.c**2 + q.d**2)
+        a = exp(q.a) * cos(vector_norm)
+        b = exp(q.a) * sin(vector_norm) * q.b / vector_norm
+        c = exp(q.a) * sin(vector_norm) * q.c / vector_norm
+        d = exp(q.a) * sin(vector_norm) * q.d / vector_norm
+
+        return Quaternion(a, b, c, d)
+
+    def ln(self):
```

Comment:
This needs a better name, ideally we would like to call `log(q)`, but that probably requires some hacks. Maybe let's keep it private for the time (prepend a `_` sign)?